core/opendaq/signal/include/opendaq/data_packet_impl.h

```diff
@@ -292,111 +295,188 @@ bool DataPacketImpl<TInterface>::isDataEqual(const DataPacketPtr& dataPacket) co
-            break;
+            const auto data = static_cast<int64_t*>(addr);
+            return Range(data[0], data[1]);
         }
         case SampleType::ComplexFloat32:
         {
-            auto data = static_cast<float*>(addr);
-            *value = ComplexNumber(data[idx * 2], data[idx * 2 + 1]).detach();
-            break;
+            const auto data = static_cast<float*>(addr);
+            return ComplexNumber(data[0], data[1]);
         }
         case SampleType::ComplexFloat64:
         {
-            auto data = static_cast<double*>(addr);
-            *value = ComplexNumber(data[idx * 2], data[idx * 2 + 1]).detach();
-            break;
+            const auto data = static_cast<double*>(addr);
+            return ComplexNumber(data[0], data[1]);
         }
         default:
         {
-            return OPENDAQ_IGNORED;
+            return BaseObject();
+        }
+    }
+}
+
+// TODO Delete this?
+template <typename TInterface>
+inline TypePtr DataPacketImpl<TInterface>::createTypeFromDescriptor(const DataDescriptorPtr& descriptor) const
+{
+    const auto fields = descriptor.getStructFields();
+    auto fieldNames = List<IString>();
+    auto fieldTypes = List<IType>();
+
+    for (auto const& field : fields)
+    {
+        TypePtr type;
+        switch (field.getSampleType())
+        {
+            case SampleType::Float32:
+            case SampleType::Float64:
+                type = SimpleType(CoreType::ctFloat);
+                break;
+            case SampleType::Int8:
+            case SampleType::UInt8:
+            case SampleType::Int16:
+            case SampleType::UInt16:
+            case SampleType::Int32:
+            case SampleType::UInt32:
+            case SampleType::Int64:
+            case SampleType::UInt64:
+                type = SimpleType(CoreType::ctInt);
+                break;
+            case SampleType::ComplexFloat32:
+            case SampleType::ComplexFloat64:
+                type = SimpleType(CoreType::ctComplexNumber);
+                break;
+            case SampleType::Struct:
+                // Recursion
+                type = createTypeFromDescriptor(field);
+                break;
+            default:
+                type = SimpleType(CoreType::ctUndefined);
+                // TODO support string, list? + test
+        }
+        fieldNames.pushBack(field.getName());
+        fieldTypes.pushBack(type);
+    }
+    return StructType(descriptor.getName(), fieldNames, fieldTypes);
+}
+
+template <typename TInterface>
+inline StructPtr DataPacketImpl<TInterface>::buildStructFromPacket(void*& addr,
+                                                                   const DataDescriptorPtr& descriptor,
+                                                                   const TypeManagerPtr& typeManager) const
+{
+    const StructTypePtr structType = createTypeFromDescriptor(descriptor);
+    typeManager.addType(structType);
```

Comment:
What happens if the type is already registered? Won't it crash?

Reply:
This code should be removed, as the type is to be registered on `setDescriptor` in SignalImpl.